src/System.Private.CoreLib/shared/System/Exception.cs

```diff
@@ -0,0 +1,199 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Runtime.Serialization;
+
+namespace System
+{
+    [Serializable]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    public partial class Exception : ISerializable
+    {
+        public Exception()
+        {
+            HResult = HResults.COR_E_EXCEPTION;
+        }
+
+        public Exception(string message)
+            : this()
+        {
+            _message = message;
+        }
+
+        // Creates a new Exception.  All derived classes should 
+        // provide this constructor.
+        // Note: the stack trace is not started until the exception 
+        // is thrown
+        // 
+        public Exception(string message, Exception innerException)
+            : this()
+        {
+            _message = message;
+            _innerException = innerException;
+        }
+
+        protected Exception(SerializationInfo info, StreamingContext context)
+        {
+            if (info == null)
+                throw new ArgumentNullException(nameof(info));
+
+            _className = info.GetString("ClassName"); // Do not rename (binary serialization)
+            _message = info.GetString("Message"); // Do not rename (binary serialization)
+            _data = (IDictionary)(info.GetValueNoThrow("Data", typeof(IDictionary))); // Do not rename (binary serialization)
+            _innerException = (Exception)(info.GetValue("InnerException", typeof(Exception))); // Do not rename (binary serialization)
+            _helpURL = info.GetString("HelpURL"); // Do not rename (binary serialization)
+            _stackTraceString = info.GetString("StackTraceString"); // Do not rename (binary serialization)
+            _HResult = info.GetInt32("HResult"); // Do not rename (binary serialization)
+            _source = info.GetString("Source"); // Do not rename (binary serialization)
+
+            RestoreRemoteStackTrace(info, context);
+        }
+
+        public virtual string Message
+        {
+            get
+            {
+                if (_message == null)
+                {
+                    return SR.Format(SR.Exception_WasThrown, GetClassName());
+                }
+                else
+                {
+                    return _message;
+                }
+            }
+        }
+
+        public virtual IDictionary Data
+        {
+            get
+            {
+                if (_data == null)
+                    _data = CreateDataContainer();
+
+                return _data;
+            }
+        }
+
+        private string GetClassName()
+        {
+            // Will include namespace but not full instantiation and assembly name.
+            if (_className == null)
```

Comment:
`Type.ToString()` returns cached string. Redundant caching here does not look worth it. (CoreRT had this cleaned up.)